builtin/providers/aws/resource_aws_instance.go

```diff
@@ -446,14 +475,23 @@ func resourceAwsInstanceRead(d *schema.ResourceData, meta interface{}) error {
 
 	// Build up the security groups
 	sgs := make([]string, len(instance.SecurityGroups))
-	for i, sg := range instance.SecurityGroups {
-		if useID {
+
+	if useID {
+		for i, sg := range instance.SecurityGroups {
 			sgs[i] = sg.Id
+		}
+		// Keep some backward compatibility. The user is warned on creation.
+		if d.Get("security_groups") != nil {
```

Comment:
This should prevent the recreation of instances in VPC when the user is still using `security_groups` in a VPC environment.

It should be removed at some stage.